src/hive/service.ts

```diff
@@ -0,0 +1,52 @@
+import { IMerjoonProjects, IMerjoonService, IMerjoonTasks, IMerjoonUsers } from '../common/types';
+import { IHiveAction, IHiveItem, IHiveProject, IHiveApi } from './types';
+import { HiveTransformer } from './transformer';
+
+export class HiveService implements IMerjoonService {
+  protected workspaceIds?: string[];
+
+  constructor(
+    public readonly api: IHiveApi,
+    public readonly transformer: HiveTransformer
+  ) {}
+
+  protected async getWorkspaceIds(): Promise<string[] | undefined> {
+    if (!this.workspaceIds) {
+      const workspaces = await this.api.v1.getWorkspaces();
+      this.workspaceIds = workspaces.map((workspace: IHiveItem) => workspace.id);
+    }
+    return this.workspaceIds;
+  }
+
+  protected async fetchAllFromWorkspaces<T>(
+    fetchFunction: (workspaceId: string) => Promise<T[]>
+  ): Promise<T[]> {
+    const workspaceIds = await this.getWorkspaceIds();
+    if (!workspaceIds) throw new Error('Missing workspaceIds');
+    
+    const items = await Promise.all(workspaceIds.map(fetchFunction));
```

Comment:
let's get rid of callbacks, try to avoid using of defining callbacks as functions as much as possible.

this time let's implement init function which gets all the workspaceId and caches them on the instance